src/jit/simdcodegenxarch.cpp

```diff
@@ -951,6 +1088,600 @@ void CodeGen::genSIMDIntrinsicUnOp(GenTreeSIMD* simdNode)
+        }
+    }
+    genProduceReg(simdNode);
+}
+
+//--------------------------------------------------------------------------------
+// genSIMDExtractUpperHalf: Generate code to extract the upper half of a SIMD register
+//
+// Arguments:
+//    simdNode - The GT_SIMD node
+//
+// Notes:
+//    This is used for the WidenHi intrinsic to extract the upper half.
+//    On SSE*, this is 8 bytes, and on AVX2 it is 16 bytes.
+//
+void CodeGen::genSIMDExtractUpperHalf(GenTreeSIMD* simdNode, regNumber srcReg, regNumber tgtReg)
+{
+    var_types simdType = simdNode->TypeGet();
+    emitAttr  emitSize = emitActualTypeSize(simdType);
+    if (compiler->getSIMDInstructionSet() == InstructionSet_AVX)
+    {
+        instruction extractIns = varTypeIsFloating(simdNode->gtSIMDBaseType) ? INS_vextractf128 : INS_vextracti128;
+        getEmitter()->emitIns_R_R_I(extractIns, EA_32BYTE, tgtReg, srcReg, 0x01);
+    }
+    else
+    {
+        instruction shiftIns = getOpForSIMDIntrinsic(SIMDIntrinsicShiftRightInternal, TYP_SIMD16);
+        if (tgtReg != srcReg)
+        {
+            inst_RV_RV(ins_Copy(simdType), tgtReg, srcReg, simdType, emitSize);
+        }
+        getEmitter()->emitIns_R_I(shiftIns, emitSize, tgtReg, 8);
+    }
+}
+
+//--------------------------------------------------------------------------------
+// genSIMDIntrinsicWiden: Generate code for SIMD Intrinsic Widen operations
+//
+// Arguments:
+//    simdNode - The GT_SIMD node
+//
+// Notes:
+//    The Widen intrinsics are broken into separate intrinsics for the two results.
+//
+void CodeGen::genSIMDIntrinsicWiden(GenTreeSIMD* simdNode)
+{
+    assert((simdNode->gtSIMDIntrinsicID == SIMDIntrinsicWidenLo) ||
+           (simdNode->gtSIMDIntrinsicID == SIMDIntrinsicWidenHi));
+
+    GenTree*  op1       = simdNode->gtGetOp1();
+    var_types baseType  = simdNode->gtSIMDBaseType;
+    regNumber targetReg = simdNode->gtRegNum;
+    assert(targetReg != REG_NA);
+    var_types      simdType = simdNode->TypeGet();
+    InstructionSet iset     = compiler->getSIMDInstructionSet();
+
+    genConsumeOperands(simdNode);
+    regNumber   op1Reg   = op1->gtRegNum;
+    regNumber   srcReg   = op1Reg;
+    emitAttr    emitSize = emitActualTypeSize(simdType);
+    instruction widenIns = getOpForSIMDIntrinsic(simdNode->gtSIMDIntrinsicID, baseType);
+
+    if (baseType == TYP_FLOAT)
+    {
+        if (simdNode->gtSIMDIntrinsicID == SIMDIntrinsicWidenHi)
+        {
+            genSIMDExtractUpperHalf(simdNode, srcReg, targetReg);
+            srcReg = targetReg;
+        }
+        inst_RV_RV(widenIns, targetReg, srcReg, simdType);
+    }
+    else
+    {
+        // targetReg = op1Reg (permute op1Reg when AVX)
+        // tmpReg = 0
+        // tmpReg = (0 > targetReg)           // (If signed) Get the sign bits
+        // punpck[l|h]dq  targetReg, tmpReg   // Interleave the sign bits
+        regNumber tmpReg = genRegNumFromMask(simdNode->gtRsvdRegs & RBM_ALLFLOAT);
```

Comment:
This comment is not correct - it needs to reflect the code being generated below.